nvdaHelper/remote/COMProxyRegistration.cpp

```diff
@@ -0,0 +1,167 @@
+	IUnknown* ClassObjPunk=NULL;
+	res=CoGetClassObject(*pProxyClsid,CLSCTX_INPROC_SERVER,nullptr,IID_IUnknown,(void**)&ClassObjPunk);
+	// From here we no longer need the activation context
+	DeactivateActCtx(0,actCtxCookie);
+	ReleaseActCtx(hActCtx);
+	if(res!=S_OK) {
+		LOG_ERROR(L"Error fetching class object for "<<dllPath<<L", code "<<res);
+		FreeLibrary(dllHandle);
+		return nullptr;
+	}
+	// Re-register the class object with COM now that the activation context is gone.
+	// Keeping the class object available to COM, with COM also handling the life time of the proxy dll now
+	DWORD dwCookie;
+	res=CoRegisterClassObject(regClsid,ClassObjPunk,CLSCTX_INPROC_SERVER,REGCLS_MULTIPLEUSE,&dwCookie);
+	ClassObjPunk->lpVtbl->Release(ClassObjPunk);
+	if(res!=S_OK) {
+		LOG_ERROR(L"Error registering class object for "<<dllPath<<L", code "<<res);
+		FreeLibrary(dllHandle);
+		return nullptr;
+	}
+	COMProxyRegistration_t* reg= new COMProxyRegistration_t();
+	reg->dllPath=dllPath;
+	reg->classObjectRegistrationCookie=dwCookie;
+	// For all interfaces the proxy dll supports, register its CLSID as their proxy stub CLSID
+	ProxyFileInfo** tempInfoPtr=pProxyInfo;
+	while(*tempInfoPtr) {
+		ProxyFileInfo& fileInfo=**tempInfoPtr;
+		for(unsigned short idx=0;idx<fileInfo.TableSize;++idx) {
+			IID iid=*(fileInfo.pStubVtblList[idx]->header.piid);
+			CLSID clsidBackup={0};
+			wstring_convert<codecvt_utf8_utf16<wchar_t>> converter;
+			wstring name=converter.from_bytes(fileInfo.pNamesArray[idx]);
+			// Fetch the old CLSID for this interface if one is set, so we can replace it on deregistration.
+			// If not set, then we'll use the standard marshaler clsid on deregistration.
+			if((res=CoGetPSClsid(iid,&clsidBackup))!=S_OK) {
+				clsidBackup=clsid_standardMarshaler;
+			} else {
+				LOG_DEBUG(L"Backed up existing clsid for interface "<<name);
+			}
+			if((res=CoRegisterPSClsid(iid,regClsid))!=S_OK) {
+				LOG_ERROR(L"Unable to register interface "<<name<<L" with proxy stub "<<dllPath<<L", code "<<res);
+				continue;
+			}
+		reg->psClsidBackups.push_back({name,iid,clsidBackup});
+		LOG_DEBUG(L"Registered interface "<<name);
+		}
+		++tempInfoPtr;
+	}
+	// We can now safely free the proxy dll. COM will keep it loaded or re-load it if needed
+	FreeLibrary(dllHandle);
+	LOG_DEBUG(L"Done registering proxy "<<dllPath);
+	return reg;
+}
+
+bool unregisterCOMProxy(COMProxyRegistration_t* reg) {
+	if(!reg) return false;
+	HRESULT res;
+	LOG_DEBUG(L"Unregistering proxy "<<(reg->dllPath));
+	for(auto& backup: reg->psClsidBackups) {
+		if((res=CoRegisterPSClsid(backup.iid,backup.clsid))!=S_OK) {
+			LOG_ERROR(L"Error registering backup PSClsid for interface "<<(backup.name)<<L" from "<<(reg->dllPath)<<L", code "<<res);
+		}
+		LOG_DEBUG(L"Unregistered interface "<<(backup.name));
+	}
+	if((res=CoRevokeClassObject((DWORD)(reg->classObjectRegistrationCookie)))!=S_OK) {
+		LOG_ERROR(L"Error unregistering class object from "<<(reg->dllPath)<<L", code "<<res);

```

Comment:
For this (and the other error case on line 156) should the function exit early, why is it ok to continue to `delete reg` and `return true`